README
========================================================================
The metar library
========================================================================

Python-metar is a python package for interpreting METAR and SPECI coded
weather reports. 

METAR and SPECI are coded aviation weather reports.  The official
coding schemes are specified in the World Meteorological Organization
(WMO) Manual on Codes, vol I.1, Part A (WMO-306 I.i.A).  US conventions
for METAR/SPECI reports vary in a number of ways from the international
standard, and are described in chapter 12 of the Federal Meteorological
Handbook No.1. (FMH-1 1995), issued by the National Oceanic and
Atmospheric Administration (NOAA).  General information about the
use and history of the METAR standard can be found at

    http://www.ncdc.noaa.gov/oa/wdc/metar/

This module extracts the data recorded in the main-body groups of
reports that follow the WMO spec or the US conventions, except for
the runway state and trend groups, which are parsed but ignored.
The most useful remark groups defined in the US spec are parsed,
as well, such as the cumulative precipitation, min/max temperature,
peak wind and sea-level pressure groups.  No other regional conventions
are formally supported, but a large number of variant formats found
in international reports are accepted.

Current METAR reports
---------------------
Current and historical METAR data can be obtained from various places.
The current METAR report for a given airport is available at the URL

    http://weather.noaa.gov/pub/data/observations/metar/stations/<station>.TXT

where <station> is the four-letter ICAO airport station code.  The 
accompanying script get_report.py will download and decode the
current report for any specified station.  

The METAR reports for all stations (worldwide) for any "cycle" (i.e., hour) 
in the last 24 hours are available in a single file at the URL

   http://weather.noaa.gov/pub/data/observations/metar/cycles/<cycle>Z.TXT

where <cycle> is a 2-digit cycle number (e.g., "00", "05" or "23").  

METAR specifications
--------------------
The Federal Meteorological Handbook No.1. (FMH-1 1995), which
describes the U.S. standards, is available online at

   http://www.ofcm.gov/fmh-1/fmh1.htm

The World Meteorological Organization (WMO) Manual on Codes, vol I.1, 
Part A (WMO-306 I.i.A) can be downloaded from site I use to distribute
this package,

   http://homepage.mac.com/wtpollard/Software/FileSharing4.html.


Tom Pollard
pollard@alum.mit.edu
May 2, 2009

------------------------------------------------------------------------
Installation
------------------------------------------------------------------------

Install this package in the usual way,

    python setup.py install

There's a small, inadequate test suite that can be run by saying

    python test/all_tests.py

There are a couple of sample scripts, described briefly below.

There's no real documentation to speak of, yet, but feel free to
contact me with any questions you might have about how to use this package.

Current sources
---------------
You can always obtain the most recent version of this package using git, via

    git clone git://github.com/tomp/python-metar.git

This is a public copy of the code repository I use for development.
Thanks to Toby White for making me aware of github.

------------------------------------------------------------------------
Contents
------------------------------------------------------------------------

README  ............... this file

parse_metar.py  ....... a simple commandline driver for the METAR parser

get_report.py ......... a script to download and decode the current reports
                        for one or more stations.

sample.py  ............ a simple script showing how the decoded data 
                        can be accessed. (see metar/*.py sources and the
                        test/test_*.py scripts for more examples.)

sample.metar  ......... a sample METAR report (longer than most).  Try
                        feeding this to the parse_metar.py script...
metar/
    Metar.py  ......... the implementation of the Metar class.  This class
                        parses and represents a single METAR report.

    Datatypes.py  ..... a support module that defines classes representing 
                        different types of meteorological data, including
                        temperature, pressure, speed, distance, direction 
                        and position.
test/
    all_tests.py  ..... a master test driver, which invokes all of the unit tests

    test_*.py  ........ individual test modules

setup.py  ............. installation script

------------------------------------------------------------------------
Example
------------------------------------------------------------------------

See the sample.py script for an annonated demonstration of the use
of this code.  Just as an appetizer, here's an interactive example...

>>> from metar import Metar
>>> obs = Metar.Metar('METAR KEWR 111851Z VRB03G19KT 2SM R04R/3000VP6000FT TSRA BR FEW015 BKN040CB BKN065 OVC200 22/22 A2987 RMK AO2 PK WND 29028/1817 WSHFT 1812 TSB05RAB22 SLP114 FRQ LTGICCCCG TS OHD AND NW -N-E MOV NE P0013 T02270215')
>>> print obs.string()
station: KEWR
type: routine report, cycle 19 (automatic report)
time: Tue Jan 11 18:51:00 2005
temperature: 22.7 C
dew point: 21.5 C
wind: variable at 3 knots, gusting to 19 knots
peak wind: WNW at 28 knots
visibility: 2 miles
visual range: runway 04R: 3000 meters to greater than 6000 meters feet
pressure: 1011.5 mb
weather: thunderstorm with rain; mist
sky: a few clouds at 1500 feet
     broken cumulonimbus at 4000 feet
     broken clouds at 6500 feet
     overcast at 20000 feet
sea-level pressure: 1011.4 mb
1-hour precipitation: 0.13in
remarks:
- Automated station (type 2)
- peak wind 28kt from 290 degrees at 18:17
- wind shift at 18:12
- frequent lightning (intracloud,cloud-to-cloud,cloud-to-ground)
- thunderstorm overhead and NW
- TSB05RAB22 -N-E MOV NE
METAR: METAR KEWR 111851Z VRB03G19KT 2SM R04R/3000VP6000FT TSRA BR FEW015 BKN040CB BKN065 OVC200 22/22 A2987 RMK AO2 PK WND 29028/1817 WSHFT 1812 TSB05RAB22 SLP114 FRQ LTGICCCCG TS OHD AND NW -N-E MOV NE P0013 T02270215
>>>>

------------------------------------------------------------------------
Tests
------------------------------------------------------------------------

The library is tested against Python 2.6, Python 2.7, and Python 3.3. A tox
(https://testrun.org/tox/latest/) configuration file is included to easily run
tests against each of these environments. To run tests against all
environments, install tox and run:

>>> tox

To run against a specific environment, use the `-e` flag:

>>> tox -e py33

------------------------------------------------------------------------
LICENSE
------------------------------------------------------------------------
The metar library is

Copyright (c) 2004-2009, Tom Pollard
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:

  Redistributions of source code must retain the above copyright
  notice, this list of conditions and the following disclaimer.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
POSSIBILITY OF SUCH DAMAGE.

------------------------------------------------------------------------


CHANGES
Change Log
==========

metar 2.0
---------
*) Updated for python 3 compatibility


metar 1.4.1
-----------
*) fixed indent levels to always be 4 spaces
*) using print() as function instead of command

metar 1.4
---------

*) Merged Toby White's changes for parsing (and ignoring) runway state groups
   and trend forecasts. (Thanks!)

*) Report the peak wind and wind shift times, if included in the remarks.
   (Requested by Daryl Herzmann.)

*) Changed the date-handling code to avoid automatic month/year adjustment
   if the month or year are specified explicitly.
   (Thanks to Scott McKuen for reporting this.)

*) Only allow single-digit numerator in fractional visibility values, to avoid
   confusion with following temp/dewpt group.
   (Reported by Franco Fiorese, Francesco Spada, and David Gregory.)

*) Fixed the misuse of str() function and string() methods in a couple of spots.
   (Reported by David Gregory and others.)

*) Added tests for the visibility, trend, and runway state parsers.

*) Added tests to verify that automatic month/year adjustment of the observation
   date were being done correctly.

*) Updated links to METAR resources in the README


metar 1.3
---------

*) The datatype classes now have __str__ methods, so you can user the
   generic str() function to get their string representation.

*) 'MMM' is now accepted to represent missing data in the wind group.

*) Fixed bugs in the handling of wind observations reported in KMH and MPH.
   (Thanks to Erik Bryer for reporting the MPH problem.)

*) '/' placeholders for missing data in the weather group are handled better now.

*) Added tests for KMH and MPH conversions to the test_speed.py script.

*) Added tests for wind-group parsing to the test_metar.py script.

*) Accept "11/2" as equivalent to "1 1/2" in the visibility group.

*) Fixed a bug that prevented the error handler in get_report.py from reporting
   the error message :-)  (Thanks to Erik Bryer for reporting this.)


metar 1.2
---------

*) Save the 1-, 3-, 6- and 24-hr preciptation values as Metar attributes.

*) Save the 6- and 24-hour max and min temperatures as Metar attributes.

*) Save the peak wind speed and direction as Metar attributes.

*) Save the sea-level pressure as a Metar attribute.  The sea-level
     pressure is treated as the station pressure, as well, if no station
     pressure is otherwise reported.

*) Added a peak_wind() method, to report peak wind data.

*) 'QNH' and 'SLP' pressure groups are now accepted.  'INS' at the end of
   a pressure group is recognized, as well.

*) A group of four digits is accepted as a pressure group, now.

*) Present-weather and cloud-cover lines are only written in the textual
   output if observations of those phenomena were reported.

*) The character 'O' is treated as a zero in the cloud-height, visibility
   and wind fields.

*) 'CORR' (in the modification field) is interpretted as 'COR'

*) 'NIL' and 'FINO' (in the modification field) are interpretted as 'NO DATA'

*) 'K', 'T' and 'LT' are all interpretted as 'KT'.

*) 'SCK' is interpretted as 'SKC'

*) 'MM' is now accepted as a placeholder in the 'temp' and 'dewpoint' groups.

*) An 'integrity' field ('+' or-) has been created.

*) 'RVRNO' groups are ignored, now.

*) '0VC' is treated as 'OVC'.

*)  A pressure group consisting of just four digits with no leading 'A' or
    'Q' is accepted now.  The units are assumed to be inches if the value
    is greater than 2500.

*) An 'O' in a cloud height field is treated as a '0'.

*) Modified to handle "MIFZFG".

*) Moved regexp matching out of the _parse* methods and into the __init__
   method.  This lets me avoid unnecessary method calls and removes a bunch
   of previously duplicated code.  It gives a 10-15% speedup.  The ability to
   ignore trend groups has been temporarily lost as a result of this change.

*) The date parsing code will assume we're in the previous month if the day
   is greater than today's date.

*) Rearranged the VISIBILITY_RE regexp to fix a parsing bug,  which had
   caused "1 1/2SM" to be parsed as two groups.

*) Modified SKY_RE regexp to accept two to four digits in the cloud height,
   and to treat 'O' as '0', if found in the height.

*) Fixed a bug (a typo) in the vertical visibility handling code.  This was
   reported by Joseph Sheedy.

*) Fixed a bug in reporting maximum visibility when there was no direction
   for the maximum visibility.

*) Changed the way the visibility group was parsed to allow better
   handling of units.  '9999U' and '7000M' are now accepted.

*) Added support for sky group of the form '///030///', indicating cloud
   height without coverage or type info.

*) Support "KTS" as an alternative to "KT".
*) Support "XX" as a placeholder for a missing temperature or dewpoint.
(These are violations of the spec, but are sometimes used, anyway.)



sample.py
#!/usr/bin/env python
#
#  This is a minimal sample script showing how the individual data
#  are accessed from the decoded report.  To produce the standard text 
#  summary of a report, the string() method of the Metar object.
#
#  The parsed data are stored as attributes of a Metar object.
#  Individual attributes are either strings. instances of one of the
#  metar.Datatypes classes, or lists of tuples of these scalars.
#  Here's a summary, adapted from the comments in the Metar.Metar.__init__()
#  method:
# 
#    Attribute       Comments [data type]
#    --------------  --------------------
#    code             original METAR code [string]
#    type             METAR (routine) or SPECI (special) [string]
#    mod              AUTO (automatic) or COR (corrected) [string]
#    station_id       4-character ICAO station code [string]
#    time             observation time [datetime]
#    cycle            observation cycle (0-23) [int]
#    wind_dir         wind direction [direction]
#    wind_speed       wind speed [speed]
#    wind_gust        wind gust speed [speed]
#    wind_dir_from    beginning of range for win dir [direction]
#    wind_dir_to      end of range for wind dir [direction]
#    vis              visibility [distance]
#    vis_dir          visibility direction [direction]
#    max_vis          visibility [distance]
#    max_vis_dir      visibility direction [direction]
#    temp             temperature (C) [temperature]
#    dewpt            dew point (C) [temperature]
#    press            barometric pressure [pressure]
#    runway           runway visibility [list of tuples...]
#                        name [string]
#                        low  [distance]
#                        high [distance]
#    weather          present weather [list of tuples...]
#                        intensity     [string]
#                        description   [string]
#                        precipitation [string]
#                        obscuration   [string]
#                        other         [string]
#    recent           recent weather [list of tuples...]
#    sky              sky conditions [list of tuples...]
#                        cover   [string]
#                        height  [distance]
#                        cloud   [string]
#    windshear        runways w/ wind shear [list of strings]
#
#    press_sea_level  sea-level pressure [pressure]
#    wind_speed_peak  peak wind speed in last hour [speed]
#    wind_dir_peak    direction of peak wind speed in last hour [direction]
#    max_temp_6hr     max temp in last 6 hours [temperature]
#    min_temp_6hr     min temp in last 6 hours [temperature]
#    max_temp_24hr    max temp in last 24 hours [temperature]
#    min_temp_24hr    min temp in last 24 hours [temperature]
#    precip_1hr       precipitation over the last hour [precipitation]
#    precip_3hr       precipitation over the last 3 hours [precipitation]
#    precip_6hr       precipitation over the last 6 hours [precipitation]
#    precip_24hr      precipitation over the last 24 hours [precipitation]
#
#    _remarks         remarks [list of strings]
#    _unparsed        unparsed remarks [list of strings]
#
#  The metar.Datatypes classes (temperature, pressure, precipitation,
#  speed, direction) describe an observation and its units.  They provide
#  value() and string() methods to that return numerical and string
#  representations of the data in any of a number of supported units.  
# 
#  (You're going to have to study the source code for more details,
#  like the available methods and supported unit conversions for the
#  metar.Datatypes objects, etc..)  

#  In particular, look at the Metar.string()
#  method, and the functions it calls.  
#
#  Feb 4, 2005 
#  Tom Pollard 
#
from metar import Metar

# A sample METAR report
code = "METAR KEWR 111851Z VRB03G19KT 2SM R04R/3000VP6000FT TSRA BR FEW015 BKN040CB BKN065 OVC200 22/22 A2987 RMK AO2 PK WND 29028/1817 WSHFT 1812 TSB05RAB22 SLP114 FRQ LTGICCCCG TS OHD AND NW-N-E MOV NE P0013 T02270215"

print("-----------------------------------------------------------------------")
print("METAR: %s" % code)
print("-----------------------------------------------------------------------")

# Initialize a Metar object with the coded report
obs = Metar.Metar(code)

# Print the individual data

# The 'station_id' attribute is a string.
print("station: %s" % obs.station_id)

if obs.type:
    print("type: %s" % obs.report_type())

# The 'time' attribute is a datetime object
if obs.time:
    print("time: %s" % obs.time.ctime())

# The 'temp' and 'dewpt' attributes are temperature objects
if obs.temp:
    print("temperature: %s" % obs.temp.string("C"))

if obs.dewpt:
      print("dew point: %s" % obs.dewpt.string("C"))

# The wind() method returns a string describing wind observations
# which may include speed, direction, variability and gusts.
if obs.wind_speed:
    print("wind: %s" % obs.wind())

# The peak_wind() method returns a string describing the peak wind 
# speed and direction.
if obs.wind_speed_peak:
    print("wind: %s" % obs.peak_wind())

# The visibility() method summarizes the visibility observation.
if obs.vis:
    print("visibility: %s" % obs.visibility())

# The runway_visual_range() method summarizes the runway visibility
# observations.
if obs.runway:
    print("visual range: %s" % obs.runway_visual_range())

# The 'press' attribute is a pressure object.
if obs.press:
    print("pressure: %s" % obs.press.string("mb"))

# The 'precip_1hr' attribute is a precipitation object.
if obs.precip_1hr:
    print("precipitation: %s" % obs.precip_1hr.string("in"))

# The present_weather() method summarizes the weather description (rain, etc.)
print("weather: %s" % obs.present_weather())

# The sky_conditions() method summarizes the cloud-cover observations.
print("sky: %s" % obs.sky_conditions("\n     "))

# The remarks() method describes the remark groups that were parsed, but 
# are not available directly as Metar attributes.  The precipitation, 
# min/max temperature and peak wind remarks, for instance, are stored as
# attributes and won't be listed here.
if obs._remarks:
    print("remarks:")
    print("- "+obs.remarks("\n- "))

print("-----------------------------------------------------------------------\n")
